Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,404 @@
+<?php
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+namespace Google\Cloud\Firestore;
+
+use Google\Cloud\Core\Exception\ServiceException;
+use SessionHandlerInterface;
+use Google\Cloud\Firestore\Connection\ConnectionInterface;
+
+/**
+ * Custom session handler backed by Cloud Firestore.
+ *
+ * Instead of storing the session data in a local file, this handler stores the
+ * data in Firestore. The biggest benefit of doing this is the data can be
+ * shared by multiple instances, making it suitable for cloud applications.
+ *
+ * The downside of using Firestore is that write operations will cost you some
+ * money, so it is highly recommended to minimize the write operations while
+ * using this handler. In order to do so, keep the data in the session as
+ * limited as possible; for example, it is ok to put only signed-in state and
+ * the user id in the session with this handler. However, for example, it is
+ * definitely not recommended that you store your application's whole undo
+ * history in the session, because every user operation will cause a Firestore
+ * write, potentially costing you a lot of money.
+ *
+ * This handler doesn't provide pessimistic lock for session data. Instead, it
+ * uses a Firestore transaction for data consistency. This means that if
+ * multiple requests are modifying the same session data simultaneously, there
+ * will be more probablity that some of the `write` operations will fail.
+ *
+ * If you are building an AJAX application which may issue multiple requests
+ * to the server, please design your session data carefully in order to avoid
+ * possible data contentions. Also please see the 2nd example below for how to
+ * properly handle errors on `write` operations.
+ *
+ * The handler stores data in a collection provided by the value of
+ * session.save_path, isolating the session data from your application data. It
+ * creates documents in the specified collection where the session name and ID
```

Comment:
I believe this is incorrect based on recent changes. Collection is a concatenation of save path and session name, and document id is session id.